### PR TITLE
[Merged by Bors] - feat: syntax linter for `#`-commands

### PR DESCRIPTION
**Pull request by adomani**
A Lean-linter implementation of the `#`-command linter introduced in #10809.

This linter produces a warning whenever a non-allowed `#`-command is used.  The allowed commands are
```lean
#align
#align_import
#noalign
#adaptation_note
```

[Zulip discussion](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.2311019.3A.20a.20syntax.20linter.20for.20.23-commands)

---

~~This PR is accompanied by #11028, which places the linter in `Tactic.Basic`, to make sure that (most of) `Mathlib` passes the `#`-command linter~~.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/HashCommandLinter.lean
Comment (alexjbest) [inline Mathlib/Tactic/HashCommandLinter.lean:23]: I don't think this should be enabled by default, but it should be enabled in CI (can do so in this PR). What do you think?

Comment (adomani) [reply Mathlib/Tactic/HashCommandLinter.lean:23]: I actually think that I prefer it to be disabled by default outside of CI: if I am using `#eval` in a file, I expect to want to see the output and not be bugged by the linter!  I'll look up how to make sure that CI uses a different default.

Comment (alexjbest) [reply Mathlib/Tactic/HashCommandLinter.lean:23]: Yeah I agree exactly, I think its enoguh to call lean in CI with `-Dlinter.hashCommand=true`

**Comment by alexjbest** [comment]
I like this much more than the sed version!

File: Mathlib/Tactic/HashCommandLinter.lean
Comment (adomani) [reply Mathlib/Tactic/HashCommandLinter.lean:23]: I am failing to make this work locally:
```
$ lake build Mathlib.mwe
[303/304] Building Mathlib.mwe
error: > LEAN_PATH=./.lake/packages/std/.lake/build/lib:./.lake/packages/Qq/.lake/build/lib:./.lake/packages/aesop/.lake/build/lib:./.lake/packages/proofwidgets/.lake/build/lib:./.lake/packages/Cli/.lake/build/lib:./.lake/packages/importGraph/.lake/build/lib:./.lake/build/lib LD_LIBRARY_PATH=./.lake/build/lib /home/damiano/.elan/toolchains/leanprover--lean4---v4.6.0-rc1/bin/lean -Dlinter.hashCommand=true -Dpp.unicode.fun=true -Dpp.proofs.withType=false -DautoImplicit=false -DrelaxedAutoImplicit=false ./././Mathlib/mwe.lean -R ././. -o ./.lake/build/lib/Mathlib/mwe.olean -i ./.lake/build/lib/Mathlib/mwe.ilean -c ./.lake/build/ir/Mathlib/mwe.c
error: stderr:
invalid -D parameter, unknown configuration option 'linter.hashCommand'
error: external command `/home/damiano/.elan/toolchains/leanprover--lean4---v4.6.0-rc1/bin/lean` exited with code 1
```

Comment (adomani) [reply Mathlib/Tactic/HashCommandLinter.lean:23]: I modified this
```lean
  weakLeanArgs :=
    if get_config? CI |>.isSome then
      #["-DwarningAsError=true", "-Dlinter.hashCommand=true"]
    else
      #["-Dlinter.hashCommand=true"]
```

Comment (adomani) [reply Mathlib/Tactic/HashCommandLinter.lean]: Alex, it seems that it is not an option to make this linter quiet locally, but chatty on CI.  I propose to make the linter available with quiet default: hopefully in the future it will be possible to set different defaults globally and then the linter would reach its full potential!

File: Mathlib/Tactic/HashCommandLinter.lean
Comment (thorimur) [inline Mathlib/Tactic/HashCommandLinter.lean:45]: We should also discard syntax like `variable ... in` and `open ... in` before linting. We can do that with something like the following, which is mostly cargo-culted from `withSetOptionIn`, but splits on the nodekind of the thing before `in`. (Not tested in this file, since my computer is slow and switching branches in VS code right now is a hassle! :P Definitely deserves tests.)

```suggestion

open Command in
/-- Exactly like `withSetOptionIn`, but recursively discards any other use of `in`. -/
private partial def withSetOptionIn' (cmd : CommandElab) : CommandElab := fun stx => do
  if stx.getKind == ``Lean.Parser.Command.in then
    if stx[0].getKind == ``Lean.Parser.Command.set_option then
      let opts ← Elab.elabSetOption stx[0][1] stx[0][2]
      Command.withScope (fun scope => { scope with opts }) do
        withSetOptionIn' cmd stx[1]
    else
      withSetOptionIn' cmd stx[1]
  else
    cmd stx

/-- Checks that no command beginning with `#` is present in 'Mathlib', except for the ones in
`whiteList`. -/
def hashCommandLinter : Linter where run := withSetOptionIn' fun stx => do
```

**Review by thorimur — COMMENTED** [review]
This review seems bigger than it is. It's basically:
* style/documentation
* code tweaking/taking care of edge cases
* questions about a couple of details

This PR looks very nice! :)

File: Mathlib/Tactic/HashCommandLinter.lean
Comment (adomani) [reply Mathlib/Tactic/HashCommandLinter.lean:48]: Removed!

File: Mathlib/Tactic/HashCommandLinter.lean
Comment (adomani) [reply Mathlib/Tactic/HashCommandLinter.lean:35]: This is a remnant of an earlier version of the linter, where I would list all atoms and scan for the ones beginning with `#` to try and do the "burrowing" that you mention above.  This is also why I later excluded spaces, since I wanted to discard matching on `# Section` in doc-modules.  I also removed the check for spaces below.

File: Mathlib/Tactic/HashCommandLinter.lean
Comment (adomani) [reply Mathlib/Tactic/HashCommandLinter.lean:52]: Removed, see above for an explanation.

**Review by adomani — COMMENTED** [review]
Thomas, as usual, thank you very much for your thorough review!  I accepted/adapted your suggestions!

I am still digesting the `withSetOptionIn'` suggestion and will report back on that!

File: Mathlib/Tactic/HashCommandLinter.lean
Comment (adomani) [reply Mathlib/Tactic/HashCommandLinter.lean]: I have not been able to get this to work.  As is, I think that it does not really enter the various `in`s.  Also, `#guard_msgs` should really be linter against anyway.

I wonder if I should follow more closely the string-based linter:
* extract the commands that begin with `#` from the environment;
* scan the syntax of the command to be linted for nodes that match the `#`-commands;
* warn on a match found.

Comment (thorimur) [reply Mathlib/Tactic/HashCommandLinter.lean]: Many apologies for the delay here—I didn't see this comment and thought you were still digesting it! 😅 (And for the past few days I've been a bit busy.)

There's a good reason my suggestion didn't work: it's wrong! :) I cargo-culted from `withSetOptionIn`, and there's a bug in `withSetOptionIn`. (leanprover/lean4#3806) Namely, it should recurse into `stx[2]` (the command syntax), not `stx[1]` (which is the token `"in"`). You'll notice that currently,
```lean
set_option lean.hashCommand true in
#print 0
```
doesn't work as expected.

So it should actually be 
```lean
open Command in
/-- Exactly like `withSetOptionIn`, but recursively discards any other use of `in`. Intended to be 
used in the hashcommand linter, where we don't need to handle non-`set_option` `in` commands. -/
private partial def withSetOptionIn' (cmd : CommandElab) : CommandElab := fun stx => do
  if stx.getKind == ``Lean.Parser.Command.in then
    if stx[0].getKind == ``Lean.Parser.Command.set_option then
      let opts ← Elab.elabSetOption stx[0][1] stx[0][2]
      Command.withScope (fun scope => { scope with opts }) do
        withSetOptionIn' cmd stx[2]
    else
      withSetOptionIn' cmd stx[2]
  else
    cmd stx
```
Note also that since `#guard_msgs in` doesn't actually use the `Lean.Parser.Command.in` parser for `in` (its `" in"` is defined in its own syntax, unlike e.g. `set_option ... in`), it's exempt from `withSetOptionIn'`. (This is also why the linter doesn't fire twice in each `#guard_msgs` test: at the top level, it hits the docstring on `#guard_msgs` and stops, which is actually quite useful.)

A good test for `withSetOptionIn'` would be something like
```lean
variable [Nonempty x] in
#synth Nonempty x
```

Since the only reason this linter doesn't catch `#guard_msgs` is the presence of a preceding docstring, and since I'm pretty sure that's unique to `#guard_msgs`, I'm okay with `#guard_msgs` not being linted against! Linting against `#guard_msgs` would also make testing the linter itself difficult. Especially since an accidental `#guard_msgs` would be (I think!) very rare.

I'm happy to push these changes if it would be convenient for you (I've now tested them locally), but if you'd like to make them yourself that's fine too! :)

File: Mathlib/Tactic/HashCommandLinter.lean
Comment (thorimur) [inline Mathlib/Tactic/HashCommandLinter.lean:48]:
```suggestion
    | some sa =>
      let a := sa.getAtomVal
      if ("#".isPrefixOf a && whitelist.all (· != a)) then
        logWarningAt sa f!"`#`-commands, such as '{a}', are not allowed in 'Mathlib'\n\
          [linter.hashCommand]"
    | none => return
```

File: Mathlib/Tactic.lean
Comment (thorimur) [inline Mathlib/Tactic.lean:80]: I wonder if we should import this in `Mathlib.Tactic.Common`, since it uses no theory files, to get it earlier in the hierarchy? What's the reasoning behind importing it here, and/or do you know if this has been discussed anywhere?

I'm also wondering if the file itself should be in new folders: Mathlib > Command > Linter > HashCommand.lean. It's a bit odd for me to see a linter in `Tactic`, but `GuardMsgs` was in `Std.Tactic`, so there is precedent for command-related things being in `Tactic`...though I think it might be nice to start having linters in their own folder one way or another! I'd like to start the discussion on zulip unless it's been discussed already.

But! I won't hold up this PR on this basis. We can always move things around later. :)

File: Mathlib/Tactic/HashCommandLinter.lean
Comment (thorimur) [reply Mathlib/Tactic/HashCommandLinter.lean]: Hmm, it's a hack to work around it now, but: what if we made the `defValue := true`, yet *also* checked `warningAsError` in the linter (and left a very visible note, and even made the description ``"enables the `#`-command linter as long as `warningAsError` is `true`"``)? I.e. `getLinterHash (← getOptions) && warningAsError.get (← getOptions) && ...`. Then it run in CI (since that uses `warningAsError := true`) but not locally under normal circumstances.

I guess the question is: in practice, would someone want to turn on this linter locally? Or is it ok if it's only ever run in CI?

(Another option is to keep `defValue := false` but turn it on whenever `warningAsError := true`.)

**Review by thorimur — COMMENTED** [review]
Apologies for the delay—very close to looking good to me! :) (Also see comment about `withSetOptionIn'` which is not organized as part of this review.)

I would like to see if we can run this during CI if possible by hijacking the `warningAsError` option (only until the underlying issue is fixed, of course)—see comment.

File: Mathlib/Tactic/HashCommandLinter.lean
Comment (adomani) [reply Mathlib/Tactic/HashCommandLinter.lean]: For me, the main use of this linter while working on a PR, would be when a `#`-command is silent.  If I am using a "noisy" `#`-command, I am probably curious about its output and adding the linter warning would be annoying.

If there were a way of making the linter noisy exactly when the `#`-command is not, then I would like it always on, I think.

**Comment by adomani** [comment]
Thomas, thank you very much for your comments!  The new `withSetOptionIn'` command works great, thanks!  I added it and also put some tests in to ensure that it does what it should!

File: Mathlib/Tactic.lean
Comment (adomani) [reply Mathlib/Tactic.lean:83]: I already have an accompanying PR where I import the linter in `Mathlib.Tactic.Basic`, since I wanted to test the linter on (most of) Mathlib.

As for moving linters to a separate folder, I think that this is a good idea, but maybe I would leave that for a second PR: there is also a `Tactic.Lint` file that should probably move there as well.

Comment (adomani) [reply Mathlib/Tactic.lean:83]: The PR with this linter imported early should be #11028.

**Comment by adomani** [comment]
@thorimur I thought that I had written this already somewhere, but maybe it got lost.

I am going to change the behaviour of the linter: it will emit a message *if and only if* the `#`-command emits no message.

This means that by leaving it always on, you can still try out a `#`-command and see its output without clutter from the linter.  However, every `#`-command now emits a message: either because the command itself does, or because the linter does (but not both!).  This means that CI will pick up in the `#`-command no matter what.

**Comment by thorimur** [comment]
Hmm, I'm not *totally* sure about only linting the silent commands, because it's nice to have an error telling you explicitly what's wrong! But it definitely works. (The "noisy stdout" section of build tells you *where* the noise is happening, right?) I'd like someone else to sign off on that specific choice, since I'm not sure what the best option is there. (Is there a discussion somewhere I might have missed?)

File: Mathlib/Tactic/HashCommandLinter.lean
Comment (thorimur) [inline Mathlib/Tactic/HashCommandLinter.lean:55]: I hope `Std.Lean.HashSet` gets upstreamed, because then you could write `{ "#align", "#align_import", "#noalign" }` without adding an import! :) Up to you if you want to prioritize fewer imports or shorter syntax, just wanted to mention it.

File: Mathlib/Tactic/HashCommandLinter.lean
Comment (thorimur) [inline Mathlib/Tactic/HashCommandLinter.lean:66]: Did the module check get lost in the mix here, or is this intentional?
```suggestion
  if getLinterHash (← getOptions) && (← get).messages.msgs.size == 0 &&
     (currentModule.getRoot == `Mathlib || currentModule == `test.HashCommandLinter) then
```

File: Mathlib/Tactic.lean
Comment (thorimur) [reply Mathlib/Tactic.lean:83]: Sounds good! :)

File: Mathlib/Tactic/HashCommandLinter.lean
Comment (thorimur) [reply Mathlib/Tactic/HashCommandLinter.lean:25]: Having it on while editing when there are no messages is a good idea! I do wonder whether it should *also* be on everywhere during CI for good error messages (as mentioned elsewhere). I guess the logic would be `getLinterHash (<- getOptions) && ((<- get).messages.msgs.size == 0 || warningAsError.get (<- getOptions))` in that case. I won't insist on this, though! :)

File: Mathlib/Tactic/HashCommandLinter.lean
Comment (adomani) [reply Mathlib/Tactic/HashCommandLinter.lean:55]: I like the notation and was not aware of it!  I added the import, since this linter is going into Mathlib, I take a nonchalant stance to "assuming" `Core` and `Std` files.

File: Mathlib/Tactic/HashCommandLinter.lean
Comment (adomani) [reply Mathlib/Tactic/HashCommandLinter.lean:66]: I intentionally made the linter check everything, including *all* `test` files.  I am not completely convinced by the change, but now you *can* unlint test files, whereas before you did not have the option to lint them, no matter what.

The PR which really makes the linter active (#11028) needed some relatively minor adjustment, but maybe I would leave the discussion of whether `test`s are in scope for the linter to that PR?

**Comment by adomani** [comment]
I [asked on Zulip](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.2311019.3A.20a.20syntax.20linter.20for.20.23-commands) about emitting an extra warning in CI, but I feel like I failed to communicate properly what the issue was.

File: Mathlib/Tactic/HashCommandLinter.lean
Comment (adomani) [reply Mathlib/Tactic/HashCommandLinter.lean]: I removed the indent, eventually!

File: Mathlib/Tactic/HashCommandLinter.lean
Comment (grunweg) [inline Mathlib/Tactic/HashCommandLinter.lean:54]: This might be clearer as `allowed_commands`, like
```lean
/-- Array of `#`-commands which are allowed in `Mathlib`. -/
private abbrev allowed_commands : HashSet String :=
```

File: Mathlib/Tactic/HashCommandLinter.lean
Comment (grunweg) [inline Mathlib/Tactic/HashCommandLinter.lean:58]: ... then this could be "for the explicitly allowed ones".

File: Mathlib/Tactic/HashCommandLinter.lean
Comment (grunweg) [inline Mathlib/Tactic/HashCommandLinter.lean:17]: Long-term, I'd really appreciate if all Lean and mathlib linters had an explanation "why is this [code pattern what we're linting] bad". For instance, in Rust, the clippy linter has these ([example](https://rust-lang.github.io/rust-clippy/master/index.html#/absolute_paths), and I found them *really* helpful.

Not strictly required, but: do you mind adding one or two sentences on why we want this?

**Review by grunweg — APPROVED** [review]
Thank you for doing this. I only have minor comments.

**Comment by adomani** [comment]
@grunweg I should have fixed all the issues.  Since your review, the changes have been to
* allow `#adaptation_note` and
* fix a syntax bug with handling nested `set_option ... in` commands.

I also extended the list of tests and went ahead to lint most of mathlib!

**Comment by grunweg** [comment]
Thanks for the quick response. I was wondering about #guard_msg; good catch!
Looking at the past logs, it seems #guard (in `test/vec_notation`) might also need to be allowed.

From my side, once CI passes, it's ready to be merged.

**Comment by adomani** [comment]
Michael, thank you for the reviews!

Re `#guard_msgs`: I simply silenced it, since it was quick and I do not expect lots of them to show up in `Mathlib`.  Probably a better solution would be to allow them only in `test` files, but I would leave this discussion for later.

Re `#guard`: I am happy not allowing it in `Mathlib`.  If I remember correctly, I actually removed some `#guard`s in `Mathlib` that were picked up by the linter and moved them to `test` files.

**Review by adomani — COMMENTED** [review]
Oh, I thought that I had replied to those previous messages, but they got lost in review!

Anyway, I ended up allowing more `#guards`!

**Comment by adomani** [comment]
Ok, after yet another failed test, I decided to make the `#guard`s not allowed, except in `test` other than `test.HashCommandLinter`.  I hope that this time the tests pass!

**Comment by grunweg** [comment]
Tests pass; I took one last look and pushed a few tiny tweaks. (@adomani Feel free to revert those if you prefer!)
I think this lint is ready to go: performance is good, mathlib passes, no bad test changes or anything that looks questionable. Thank you for your work on this, this one took a while - but I'm glad it lands!
maintainer merge

(A follow-up PR could remove the sed-based linter this replaces. Imho, this doesn't have to happen in this PR.)

**Comment by grunweg** [comment]
Filed #13142 for the follow-up.

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**
